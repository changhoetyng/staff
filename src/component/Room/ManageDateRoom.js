import React, { Component } from "react";
import Header from "../../sharedComponent/Header";
import Table from "react-bootstrap/Table";
import DropdownButton from "react-bootstrap/DropdownButton";
import Dropdown from "react-bootstrap/Dropdown";
import DatePicker from "react-datepicker";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import "react-datepicker/dist/react-datepicker.css";
import moment from "moment";
import Button from "react-bootstrap/Button";
import Modal from "react-bootstrap/Modal";
import { api } from "../../api/api";

class ManageDateRoom extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      currentRoomId: null,
      selectedSub: null,
      selectedSubId: null,
      date: null,
      checkedItems: new Map(),
      activeData: { time: "", timeStatus: { status: "", studentId: "" } },
      modalOn: false,
      data: [],
    };
  }

  async componentDidMount() {
    this.setState({ loading: true });
    await api
      .get("/room/getRoom")
      .then((res) => {
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  dropdownSelection(selected, id) {
    this.setState({ currentSelection: selected, currentRoomId: id });
  }

  dropdownSub(selected, id) {
    this.setState({ selectedSub: selected, selectedSubId: id });
  }

  renderRoomButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.currentSelection
            ? this.state.currentSelection
            : "Select a Room"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            <Dropdown.Item
              key={i}
              as="button"
              onSelect={() => this.dropdownSelection(v.room, v.roomId)}
            >
              {v.room}
            </Dropdown.Item>
          );
        })}
      </DropdownButton>
    );
  }

  renderCategoryButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          !this.state.currentSelection
            ? "Please Select a Room"
            : this.state.selectedSub
            ? this.state.selectedSub
            : "Select a subcategory"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            v.roomId === this.state.currentRoomId && v.subCategory &&
                v.subCategory.map((sub,j) => {
                  return (
                    <Dropdown.Item
                      key={j}
                      as="button"
                      onSelect={() =>
                        this.dropdownSub(sub.subName, sub._id)
                      }
                    >
                      {sub.subName}
                    </Dropdown.Item>
                  );
                })
            
          );
        })}
      </DropdownButton>
    );
  }

  handleChange = (e) => {
    const item = e.target.name;
    const isChecked = e.target.checked;
    this.setState((prevState) => ({
      checkedItems: prevState.checkedItems.set(item, isChecked),
    }));
  };

  renderTableContent() {
    let selectedRoom = this.state.data.find(
      (v) => v.roomId === this.state.currentRoomId 
    );

    let selectedDate = selectedRoom
      ? selectedRoom.data.find(
          (v) => v.date === moment(this.state.date).format("DD/MM/YYYY") && v.subCategoryId === this.state.selectedSubId
        )
      : null;
    let timeData = selectedDate ? selectedDate.timeListing : [];
    return (
      <tbody>
        {timeData.map((v, i) => {
          return (
            <tr key={i}>
              <td>{v.time}</td>
              <td>{v.timeStatus.status}</td>
              <td>
                <Button
                  variant="link"
                  style={{ marginLeft: 10 }}
                  onClick={() =>
                    this.setState({ activeData: v, modalOn: true })
                  }
                >
                  More Details
                </Button>
              </td>
            </tr>
          );
        })}
      </tbody>
    );
  }

  hideModal() {
    this.setState({ modalOn: false });
  }

  async openBooking() {
    this.setState({ loading: true });
    await api
      .patch("/room/openTime", {
        roomId: this.state.currentRoomId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  async closeBooking() {
    this.setState({ loading: true });
    await api
      .patch("/room/closeTime", {
        roomId: this.state.currentRoomId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  async cancelBooking() {
    this.setState({ loading: true });
    await api
      .patch("/room/cancelBooking", {
        roomId: this.state.currentRoomId,
        date: moment(this.state.date).format("DD/MM/YYYY"),
        time: this.state.activeData.time,
        subCategoryId: this.state.selectedSubId
      })
      .catch((err) => {
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, modalOn: false });
  }

  render() {
    return (
      <div>
        <Header history={this.state.history} />
        <Modal
          show={this.state.modalOn}
          onHide={() => this.hideModal()}
          centered
        >
          <Modal.Header>
            <Modal.Title>Details</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            Date: {moment(this.state.date).format("DD/MM/YYYY")}
            <br />
            Time: {this.state.activeData.time}
            <br />
            Status: {this.state.activeData.timeStatus.status}
            <br />
            {this.state.activeData.timeStatus.status === "booked" && (
              <div>
                Student ID: {this.state.activeData.timeStatus.studentId}
              </div>
            )}
          </Modal.Body>
          <Modal.Footer>
            {this.state.activeData.timeStatus.status === "booked" && (
              <Button variant="danger" onClick={() => this.cancelBooking()}>
                Cancel booking
              </Button>
            )}
            {this.state.activeData.timeStatus.status === "open" && (
              <Button onClick={() => this.closeBooking()}>Close booking</Button>
            )}
            {this.state.activeData.timeStatus.status === "close" && (
              <Button onClick={() => this.openBooking()}>Open booking</Button>
            )}
          </Modal.Footer>
        </Modal>
        <div className="container float-left">
          <div className="container-fluid" style={{ marginTop: 20 }}>
            <h4>Sport Complex</h4>
          </div>
          <div
            className="container-fluid"
            style={{ marginTop: 20, flexDirection: "row" }}
          >
            <Row>
              <Col>{this.renderRoomButton()}</Col>
              <Col>{this.renderCategoryButton()}</Col>
              {
                <Col>
                  <DatePicker
                    selected={this.state.date}
                    isClearable
                    onChange={(date) => this.setState({ date })}
                  />
                </Col>
              }
            </Row>
          </div>
          <div className="container-fluid" style={{ marginTop: 20 }}>
            <Table striped bordered hover>
              <thead>
                <tr>
                  <th>Time</th>
                  <th>Status</th>
                  <th>Details</th>
                </tr>
              </thead>
              {this.renderTableContent()}
            </Table>
          </div>
        </div>
      </div>
    );
  }
}

export default ManageDateRoom;
