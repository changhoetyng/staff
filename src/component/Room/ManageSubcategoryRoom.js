import React, { Component } from "react";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import Dropdown from "react-bootstrap/Dropdown";
import DropdownButton from "react-bootstrap/DropdownButton";
import FullPageLoader from "../../hooks/FullPageLoader";
import Header from "../../sharedComponent/Header";
import "../../css/Setting.css";
import { api } from "../../api/api";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Modal from "react-bootstrap/Modal";
import { QRCode } from "react-qr-svg";

class ManageSubcategoryRoom extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      currentRoomId: null,
      subCategory: "",
      data: [],
      loading: false,
      qrCodeModal: false,
      selectedQR: "",
      isAdmin: false
    };
  }

  async componentDidMount() {
    this.setState({ loading: true });
    await api
      .get("/user")
      .then((res) => {
        const role = res.data.user.role;
        if (role === "admin") {
          this.setState({ isAdmin: true });
        }
      })
      .catch((err) => console.log(err));

    await api
      .get("/room/getRoom")
      .then((res) => {
        console.log(res.data.data);
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  dropdownSelection(selected, id) {
    this.setState({ currentSelection: selected, currentRoomId: id });
  }

  renderRoomButton() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.currentSelection
            ? this.state.currentSelection
            : "Select a Room"
        }
      >
        {this.state.data.map((v, i) => {
          return (
            <Dropdown.Item
              key={i}
              as="button"
              onSelect={() => this.dropdownSelection(v.room, v.roomId)}
            >
              {v.room}
            </Dropdown.Item>
          );
        })}
      </DropdownButton>
    );
  }

  openQrCode(sub) {
    this.setState({ selectedQR: sub, qrCodeModal: true });
  }

  handleClose() {
    this.setState({ qrCodeModal: false });
  }

  renderTableContent() {
    let selectedRoom = this.state.data.find(
      (v) => v.roomId === this.state.currentRoomId
    );
    let subCategory = selectedRoom ? selectedRoom.subCategory : [];
    return (
      <div className="row">
        {subCategory &&
          subCategory.map((sub, i) => {
            return (
              <div className="col" key={i}>
                <Card
                  style={{
                    width: "19rem",
                    whiteSpace: "pre-wrap",
                    paddingRight: 0,
                  }}
                >
                  <Card.Body>
                    <Card.Title>{sub.subName}</Card.Title>
                    <Card.Text>
                      {sub.currentUser ? (
                        <p>Current User: {sub.currentUser}</p>
                      ) : (
                        <p>Current User: Not occupied</p>
                      )}
                    </Card.Text>
                    <Card.Link
                      style={{ cursor: "pointer" }}
                      onClick={() => {
                        this.openQrCode(sub);
                      }}
                    >
                      QR Code
                    </Card.Link>
                    { this.state.isAdmin &&
                    <Card.Link
                      style={{ cursor: "pointer" }}
                      onClick={() => {
                        this.removesubcategory(sub._id);
                      }}
                    >
                      Remove
                    </Card.Link>
                    }
                  </Card.Body>
                </Card>
              </div>
            );
          })}
      </div>
    );
  }

  async removesubcategory(subId) {
    this.setState({ loading: true });

    await api
      .patch("/room/deletesub", {
        roomId: this.state.currentRoomId,
        subCategoryId: subId,
      })
      .catch((err) => {
        alert(err);
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false });
  }

  async addsubcategory() {
    this.setState({ loading: true });

    await api
      .patch("/room/addSub", {
        roomId: this.state.currentRoomId,
        subCategory: this.state.subCategory,
      })
      .catch((err) => {
        alert(err);
        console.log(err);
      });
    this.componentDidMount();
    this.setState({ loading: false, subCategory: "" });
  }

  render() {
    return (
      <div>
        {this.state.loading && <FullPageLoader />}
        <Header history={this.state.history} />
        <div className="container float-left non-printable">
          <h2 id="manageSubcategoryTitle">Manage Subcategory</h2>

          <div style={{ marginTop: 20 }}>
            <h4 id="title">Room</h4>
            <div className="container float-left">
              <div className="row" style={{ marginTop: 20 }}>
                {this.renderRoomButton()}
              </div>
              <div className="row" style={{ marginBottom: 20, marginTop: 20 }}>
                {this.renderTableContent()}
              </div>
              { this.state.isAdmin &&
              <div className="row,container float-left">
                <div className="row">
                  <p className="word">Manage Subcategory:</p>
                  <div className="col">
                    <input
                      type="text"
                      value={this.state.subCategory}
                      onChange={(e) =>
                        this.setState({ subCategory: e.target.value })
                      }
                    />
                  </div>
                  <div className="col">
                    <Button onClick={() => this.addsubcategory()}>Add</Button>
                  </div>
                </div>
              </div>
              }
            </div>
          </div>
        </div>
        <div class="non-printable">
          <Modal
            show={this.state.qrCodeModal}
            aria-labelledby="contained-modal-title-vcenter"
            onHide={() => this.handleClose()}
            centered
          >
            <Modal.Header>
              <Modal.Title>QR Code</Modal.Title>
            </Modal.Header>
            <Modal.Body>
              <div class="printable">
                <h2 style={{ textAlign: "center" }}>
                  {this.state.currentSelection}
                </h2>
                <h4 style={{ textAlign: "center" }}>
                  {this.state.selectedQR.subName}
                </h4>
                <div style={{ textAlign: "center" }}>
                  <QRCode
                    level="Q"
                    style={{ width: 256 }}
                    value={JSON.stringify({
                      venueId: this.state.currentRoomId,
                      subCategoryId: this.state.selectedQR._id,
                    })}
                  />
                </div>
              </div>
            </Modal.Body>
            <Modal.Footer>
              <Button onClick={() => window.print()}>Print</Button>
            </Modal.Footer>
          </Modal>
        </div>
      </div>
    );
  }
}

export default ManageSubcategoryRoom;
