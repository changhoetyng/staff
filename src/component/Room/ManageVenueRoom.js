import React, { Component } from "react";
import Button from "react-bootstrap/Button";
import Card from "react-bootstrap/Card";
import FullPageLoader from "../../hooks/FullPageLoader";
import DatePicker from "react-datepicker";
import Header from "../../sharedComponent/Header";
import "../../css/Setting.css";
import { api } from "../../api/api";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Dropdown from "react-bootstrap/Dropdown";
import Modal from "react-bootstrap/Modal";
import DropdownButton from "react-bootstrap/DropdownButton";
import moment from "moment"

class ManageVenue extends Component {
  constructor(props) {
    super(props);
    this.state = {
      history: this.props.history,
      currentSelection: null,
      selectedSubCategory: null,
      data: [],
      loading: false,
      fields: {},
      isModal: false,
      selectedSubCategoryId: null,
      error: ""
    };
  }

  async componentDidMount() {
    this.setState({ loading: true });
    await api
      .get("/room/getRoom")
      .then((res) => {
        this.setState({ data: res.data.data });
      })
      .catch((err) => {
        console.log(err);
      });
    this.setState({ loading: false });
  }

  handleDateChange = (dateName, dateValue) => {
    this.setState({
      fields: {
        ...this.state.fields,
        [dateName]: dateValue,
      },
    });
  };

  renderDropdown() {
    return (
      <DropdownButton
        id="dropdown-item-button"
        title={
          this.state.selectedSubCategory
            ? this.state.selectedSubCategory
            : "Select"
        }
      >
        {this.state.currentSelection &&
          this.state.currentSelection.subCategory &&
          this.state.currentSelection.subCategory.map((v, i) => {
            return (
              <Dropdown.Item
                key={i}
                as="button"
                onSelect={() =>
                  this.setState({
                    selectedSubCategoryId: v._id,
                    selectedSubCategory: v.subName,
                  })
                }
              >
                {v.subName}
              </Dropdown.Item>
            );
          })}
      </DropdownButton>
    );
  }

  renderRoomCard() {
    return this.state.data.map((data, index) => {
      return (
        <Col key={index}>
          <Card className="text-center" style={{ width: 300 }}>
            <Card.Body>
              <Card.Title>{data.room}</Card.Title>
              <DatePicker
                selected={this.state.fields[data.roomId]}
                minDate={new Date()}
                isClearable
                onChange={(date) =>
                  this.handleDateChange(data.roomId, date)
                }
              />
            </Card.Body>
            {this.state.fields[data.roomId] && (
              <Card.Body>
                <Card.Link
                  style={{ cursor: "pointer" }}
                  onClick={() =>
                    this.setState({ currentSelection: data, isModal: true })
                  }
                >
                  Open Date
                </Card.Link>
              </Card.Body>
            )}
          </Card>
        </Col>
      );
    });
  }

  async openDate() {
    const roomId = this.state.currentSelection.roomId
    const subCategoryId = this.state.selectedSubCategoryId
    const date = moment(this.state.fields[roomId]).format("DD/MM/YYYY")
    this.setState({ loading: true });
    await api
      .post("/room/openDate",{
        roomId,
        date,
        subCategoryId
      })
      .then(() => {
        this.setState({isModal: false})
      })
      .catch((err) => {
        if(err.response.data.message === "time already exists"){
          this.setState({error: "Time already existed"})
        }
      })
      this.setState({ loading: false });
  }

  handleClose() {
    this.setState({
      isModal: false,
      selectedSubCategory: null,
      selectedSubCategoryId: null,
      currentSelection: null,
      error: ""
    });
  }

  render() {
    return (
      <div>
        <Modal
          show={this.state.isModal}
          aria-labelledby="contained-modal-title-vcenter"
          onHide={() => this.handleClose()}
          centered
        >
          <Modal.Body className="text-center">
            <h4>Select subcategory</h4>
            <div>{this.renderDropdown()}</div>
            {this.state.selectedSubCategoryId && (<Button onClick={() => this.openDate()}>Open</Button>)}
            <p style={{ color: "red", alignSelf: "left" }}>
                  {this.state.error}
            </p>
          </Modal.Body>
        </Modal>
        {this.state.loading && <FullPageLoader />}
        <Header history={this.state.history} />
        <div className="container float-left">
          <h2>Manage Venue</h2>
          <div style={{ marginTop: 20 }}>
            <h4>Room</h4>
          </div>
          <div style={{ marginTop: 20 }}>
            <Row>{this.renderRoomCard()}</Row>
          </div>
        </div>
      </div>
    );
  }
}

export default ManageVenue;
