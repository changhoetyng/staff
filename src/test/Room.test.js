import React from "react";
import { configure, shallow } from "enzyme";

// Import Page
import ManageDateRoom from "../component/Room/ManageDateRoom";
import ManageSubcategoryRoom from "../component/Room/ManageSubcategoryRoom";
import ManageVenueRoom from "../component/Room/ManageVenueRoom";

describe("ManageDateRoom", () => {
  const manageDateRoom = shallow(<ManageDateRoom />);

  test("Title should contain word \"Room\"", () => {
    const title = manageDateRoom.find("#title");
    expect(title.text()).toBe("Room");
  });

  test("Snapshot ManageDateRoom", () => {
    expect(manageDateRoom).toMatchSnapshot();
  });
});

describe("ManageSubcategoryRoom", () => {
  const manageSubcategoryRoom = shallow(
    <ManageSubcategoryRoom />
  );

  test("Title should contain word \"Room\"", () => {
    const title = manageSubcategoryRoom.find("#title");
    expect(title.text()).toBe("Room");
  });

  test("manageSubcategoryTitle should contain word \"Manage Subcategory\"", () => {
    const title = manageSubcategoryRoom.find("#manageSubcategoryTitle");
    expect(title.text()).toBe("Manage Subcategory");
  });

  test("Snapshot ManageSubcategoryRoom", () => {
    expect(manageSubcategoryRoom).toMatchSnapshot();
  });
});

describe("ManageVenueRoom", () => {
  const manageVenueRoom = shallow(<ManageVenueRoom />);

  test("change password button should activate modal", () => {
    const title = manageVenueRoom.find("#title");
    expect(title.text()).toBe("Room");
  });

  test("Title should contain word \"Manage Subcategory\"", () => {
    const title = manageVenueRoom.find("#manageVenueTitle");
    expect(title.text()).toBe("Manage Venue");
  });

  test("Snapshot ManageVenueRoom", () => {
    expect(manageVenueRoom).toMatchSnapshot();
  });
});
