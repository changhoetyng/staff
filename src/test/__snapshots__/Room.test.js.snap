// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDateRoom Snapshot ManageDateRoom 1`] = `ShallowWrapper {}`;

exports[`ManageSubcategoryRoom Snapshot ManageSubcategoryRoom 1`] = `ShallowWrapper {}`;

exports[`ManageVenueRoom Snapshot ManageVenueRoom 1`] = `ShallowWrapper {}`;
